### PR TITLE
Migrates Channel.GetByName to sync by default

### LINKED ISSUE #11171: Migrate "Channel.GetByNameIncludeDeleted" to Sync by default
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/MM-16318)

Mattermost is migrating its Store layer to be sync by default, and only use Async when needed and we're looking for contributors to help with that effort. This Help Wanted issue is to migrate the `GetByNameIncludeDeleted` in the `Channel` store.

The expected way to implement it is, go to the `Channel` store implementation in the `store/sqlstore/` directory, modify the method `GetByNameIncludeDeleted` to return directly an object from the `model` module, and a `*model.AppError` (removing the `store.Do` wrapper). After that, you must modify the interface defined in `store/store.go` to match with the changes. Then, you should execute `make store-mocks` to rebuild the mocks with the new interface. And finally, modify the rest of the code (tests included) to use the new interface of the function properly.

Example: mattermost/mattermost-server#10613

### LINKED ISSUE #11173: Migrate "Channel.GetByName" to Sync by default
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/MM-16316)

Mattermost is migrating its Store layer to be sync by default, and only use Async when needed and we're looking for contributors to help with that effort. This Help Wanted issue is to migrate the `GetByName` in the `Channel` store.

The expected way to implement it is, go to the `Channel` store implementation in the `store/sqlstore/` directory, modify the method `GetByName` to return directly an object from the `model` module, and a `*model.AppError` (removing the `store.Do` wrapper). After that, you must modify the interface defined in `store/store.go` to match with the changes. Then, you should execute `make store-mocks` to rebuild the mocks with the new interface. And finally, modify the rest of the code (tests included) to use the new interface of the function properly.

Example: mattermost/mattermost-server#10613

### PR DESCRIPTION
**Pull request by rvillablanca**
#### Summary

This PR makes `Channel.GetByName` and `Channel.GetByNameIncludedDeleted` sync by default.
Closes #11171
Closes #11173

### CODE REVIEW

File: app/channel.go [hunk 1]
```diff
@@ -74,10 +74,9 @@ func (a *App) JoinDefaultChannels(teamId string, user *model.User, shouldBeAdmin
 
 	var err *model.AppError
 	for _, channelName := range a.DefaultChannelNames() {
-		if result := <-a.Srv.Store.Channel().GetByName(teamId, channelName, true); result.Err != nil {
```
Comment (jespino) [inline app/channel.go]: nit: a possible improvement here could be something like:

```go
channel, channelErr := a.Srv.Store.Channel().GetByName(teamId, channelName, true)
if channelErr != nil {
    err = channelErr
    continue
}

// ... here the rest of the code without the else
```

**Comment by jespino** [comment]
@rvillablanca some shadow errors in the govet.

**Comment by rvillablanca** [comment]
> @rvillablanca some shadow errors in the govet.

I fixed all go vet warnings but I'm a little afraid if this could change some behaviour. If jenking run all test with success, then should I expect everything works fine ?

File: app/channel.go [hunk 2]
```diff
@@ -314,7 +314,7 @@ func (a *App) GetOrCreateDirectChannel(userId, otherUserId string) (*model.Chann
 			if esInterface != nil && *a.Config().ElasticsearchSettings.EnableIndexing {
 				a.Srv.Go(func() {
 					for _, id := range []string{userId, otherUserId} {
-						if err := a.indexUserFromId(id); err != nil {
```
Comment (jespino) [inline app/channel.go]: Clearly here I wouldn't reuse the variable from the outside of the gorouting. I suggest here to rename the error variable instead.

File: app/webhook.go [hunk 3]
```diff
@@ -601,14 +601,16 @@ func (a *App) HandleIncomingWebhook(hookId string, req *model.IncomingWebhookReq
 		} else if channelName[0] == '#' {
 			cchan = make(store.StoreChannel, 1)
 			go func() {
-				channel, err := a.Srv.Store.Channel().GetByName(hook.TeamId, channelName[1:], true)
+				var err *model.AppError
+				channel, err = a.Srv.Store.Channel().GetByName(hook.TeamId, channelName[1:], true)
```
Comment (jespino) [inline app/webhook.go]: Here I wouldn't reuse the channel variable. Change the channel variable to use other name instead.

File: app/webhook.go [hunk 4]
```diff
@@ -601,14 +601,16 @@ func (a *App) HandleIncomingWebhook(hookId string, req *model.IncomingWebhookReq
...
 				cchan <- store.StoreResult{Data: channel, Err: err}
 				close(cchan)
 			}()
 		} else {
 			cchan = make(store.StoreChannel, 1)
 			go func() {
-				channel, err := a.Srv.Store.Channel().GetByName(hook.TeamId, channelName, true)
+				var err *model.AppError
+				channel, err = a.Srv.Store.Channel().GetByName(hook.TeamId, channelName, true)
```
Comment (jespino) [inline app/webhook.go]: Here I wouldn't reuse the channel variable. Change the channel variable to use other name instead.

File: app/channel.go (hunk 2 above)
Comment (rvillablanca) [reply app/channel.go]: I corrected this

File: app/webhook.go (hunk 3 above)
Comment (rvillablanca) [reply app/webhook.go]: Changed

File: app/webhook.go (hunk 4 above)
Comment (rvillablanca) [reply app/webhook.go]: Changed

**Comment by jespino** [comment]
@rvillablanca Could you fix the shadow variable?

**Comment by rvillablanca** [comment]
ok the `go vet` was corrected and the code was updated with `master` @jespino @enahum

**Pull request merged**